File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BaseSingleObjectSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.concurrent.Callable;

import eu.pretix.libpretixsync.api.ApiException;
import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.api.ResourceNotModified;
import eu.pretix.libpretixsync.db.RemoteObject;
import eu.pretix.libpretixsync.utils.JSONUtils;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;

public abstract class BaseSingleObjectSyncAdapter<T extends RemoteObject & Persistable> implements DownloadSyncAdapter {

    protected BlockingEntityStore<Persistable> store;
    protected PretixApi api;
    protected String eventSlug;
    protected String key;
    protected SyncManager.ProgressFeedback feedback;
    protected SyncManager.CanceledState canceledState;

    public BaseSingleObjectSyncAdapter(BlockingEntityStore<Persistable> store, String eventSlug, String key, PretixApi api, SyncManager.ProgressFeedback feedback) {
        this.store = store;
        this.api = api;
        this.eventSlug = eventSlug;
        this.key = key;
        this.feedback = feedback;
    }

    @Override
    public void setCancelState(SyncManager.CanceledState state) {
        canceledState = state;
    }

    @Override
    public void download() throws JSONException, ApiException {
        if (feedback != null) {
            feedback.postFeedback("Downloading " + getResourceName() + "…");
        }
        try {
            JSONObject data = downloadRawData();
            processData(data);
        } catch (ResourceNotModified e) {
            // Do nothing
        }
    }

    abstract T getKnownObject();

    protected void processData(final JSONObject data) {
        store.runInTransaction(new Callable<Void>() {
            @Override
            public Void call() throws Exception {
                T known = getKnownObject();
                T obj;
                JSONObject old = null;
                if (known != null) {
                    obj = known;
                    old = obj.getJSON();
                    if (!JSONUtils.similar(data, old)) {
                        updateObject(obj, data);
                        store.update(obj);
                    }
                } else {
                    obj = newEmptyObject();
                    updateObject(obj, data);
                    store.insert(obj);
                }
                return null;
            }
        });
    }

    public abstract void updateObject(T obj, JSONObject jsonobj) throws JSONException;

    protected String getUrl() {
        return api.eventResourceUrl(getResourceName() + "/" + key);
    }

    protected JSONObject downloadRawData() throws ApiException, ResourceNotModified {
        return downloadPage(getUrl());
    }

    protected JSONObject downloadPage(String url) throws ApiException, ResourceNotModified {
        return api.fetchResource(url).getData();
    }

    abstract String getResourceName();

    abstract T newEmptyObject();
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/BaseSingleObjectSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import eu.pretix.libpretixsync.api.ApiException
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.api.ResourceNotModified
import eu.pretix.libpretixsync.sqldelight.Migrations
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.CanceledState
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import eu.pretix.libpretixsync.utils.JSONUtils
import org.json.JSONException
import org.json.JSONObject

abstract class BaseSingleObjectSyncAdapter<T>(
    protected var db: SyncDatabase,
    protected var eventSlug: String,
    protected var key: String,
    protected var api: PretixApi,
    protected var syncCycleId: String,
    protected var feedback: ProgressFeedback? = null,
) : DownloadSyncAdapter {

    private var canceledState: CanceledState? = null

    override fun download() {
        if (feedback != null) {
            feedback!!.postFeedback("Downloading " + getResourceName() + " [" + eventSlug + "] …")
        }
        try {
            val data = downloadRawData()
            processData(data)
        } catch (e: ResourceNotModified) {
            // Do nothing
        }
    }

    override fun setCancelState(state: SyncManager.CanceledState?) {
        canceledState = state
    }

    abstract fun getKnownObject(): T?

    protected fun processData(data: JSONObject) {
        try {
            data.put("__libpretixsync_dbversion", Migrations.CURRENT_VERSION)
            data.put("__libpretixsync_syncCycleId", syncCycleId)
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        runInTransaction {
            val known = getKnownObject()
            if (known != null) {
                val old = getJSON(known)
                if (!JSONUtils.similar(data, old)) {
                    update(known, data)
                }
            } else {
                insert(data)
            }
        }
    }

    protected open fun getUrl(): String {
        return api.eventResourceUrl(eventSlug, getResourceName() + "/" + key)
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    protected fun downloadRawData(): JSONObject {
        return downloadPage(getUrl())
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    protected fun downloadPage(url: String): JSONObject {
        return api.fetchResource(url).data!!
    }

    abstract fun getResourceName(): String

    abstract fun runInTransaction(body: TransactionWithoutReturn.() -> Unit)

    abstract fun insert(jsonobj: JSONObject)

    abstract fun update(obj: T, jsonobj: JSONObject)

    abstract fun getJSON(obj: T): JSONObject
}
